compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5561,11 +5561,23 @@ private BType resolveQueryType(SymbolEnv env, BLangExpression selectExp, BType c
                     selectType = checkExpr(selectExp, env, type, data);
                     resolvedType = selectType;
                     break;
+                case TypeTags.MAP:
+                    List<BType> memberTypes = new ArrayList<>();
+                    memberTypes.add(symTable.stringType);
+                    memberTypes.add(((BMapType) type).getConstraint());
+                    BTupleType newExpType = new BTupleType(null, memberTypes);
+                    selectType = checkExpr(selectExp, env, newExpType, data);
+                    resolvedType = selectType;
+                    break;
                 case TypeTags.NONE:
                 default:
                     // contextually expected type not given (i.e var).
                     selectType = checkExpr(selectExp, env, type, data);
-                    resolvedType = getNonContextualQueryType(selectType, collectionType);
+                    if (queryExpr.isMap) {
+                        resolvedType = symTable.mapType;
```

Comment:
We have to move this check into `getNonContextualQueryType` as we did in stream and table